src/ddsrt/include/dds/ddsrt/hopscotch.h

```diff
@@ -32,42 +47,172 @@ extern "C" {
+ * 
+ * Hopscotch needs to be able to compare two elements to confirm it has found the correct one
+ * (equal hashes doesn't guarantee that the keys are equal).
+ * The input arguments are pointers to the user data objects to compare.
+ * Returns int meant to be interpreted as bool (1 is match, 0 is no match)
  */
 typedef int (*ddsrt_hh_equals_fn) (const void *a, const void *b);
 
-/*
- * Hopscotch is will resize its internal buckets list when needed. It will
+/**
+ * @brief User defined garbage collection function for buckets (only used for concurrent version).
+ * 
+ * Hopscotch will resize its internal buckets list when needed. It will
  * call this garbage collection function with the old buckets list. The
  * caller has to delete the list when it deems it safe to do so.
  */
 typedef void (*ddsrt_hh_buckets_gc_fn) (void *bs, void *arg);
 
 /* Sequential version */
+/**
+ * @brief The hopscotch hash table.
+ * @see @ref ddsrt_hh_new
+ * @see @ref ddsrt_hh_free
+ */
 struct ddsrt_hh;
 
+/**
+ * @brief Iter object for the iterator to store its progress and know where to go next.
+ * @see @ref ddsrt_hh_iter_first
+ */
 struct ddsrt_hh_iter {
     struct ddsrt_hh *hh;
     uint32_t cursor;
 };
 
+/**
+ * @brief Create a hopscotch hash table.
+ * 
+ * @param[in] init_size the minimum initial size
+ * @param[in] hash the hash function, see @ref ddsrt_hh_hash_fn
+ * @param[in] equals the equals function, see @ref ddsrt_hh_equals_fn
+ * @return pointer to the hopscotch hash table
+ * @see @ref ddsrt_hh_free
+ */
 DDS_EXPORT struct ddsrt_hh *ddsrt_hh_new (uint32_t init_size, ddsrt_hh_hash_fn hash, ddsrt_hh_equals_fn equals) ddsrt_nonnull_all;
+
+/**
+ * @brief Destroy a hopscotch hash table.
+ * 
+ * It destroys the hash table and its buckets, but not the user data in the buckets.
+ * The user data in the table must be deleted prior to calling this,
+ * which can be done by using @ref ddsrt_hh_enum with a free function.
+ * 
+ * @param[in,out] hh the hash table to destroy
+ * @see @ref ddsrt_hh_new
+ */
 DDS_EXPORT void ddsrt_hh_free (struct ddsrt_hh * __restrict hh) ddsrt_nonnull_all;
+
+/**
+ * @brief Lookup an element in the hash table.
+ * 
+ * Despite the name, the parameter 'keyobject' is a pointer to the whole user data object, not just the key,
+ * so you need a dummy object to do the lookup. However, if the user data object is defined such that the key is its first element,
+ * then it is also possible to do the lookup by making just a key object and passing a pointer to that key.
+ * 
+ * @param[in] rt the hash table
+ * @param[in] keyobject is the object with which to do the lookup
+ * @return pointer to the matching element, NULL if failed
+ */
 DDS_EXPORT void *ddsrt_hh_lookup (const struct ddsrt_hh * __restrict rt, const void * __restrict keyobject) ddsrt_nonnull_all;
+
+/**
+ * @brief Add an element to the hash table.
+ * 
+ * It is not allowed to add duplicates to the hash table. Trying to do so results in failure.
```

Comment:
I would phrase this differently, because this phrasing leaves it unclear whether you're allowed to try (you are).

Something like: attempts to store an element with a key that is already present will be rejected.

And then the return value becomes something like: 1 is success, element stored; 0 is failure, element key already present.